errorlint/lint.go

```diff
@@ -353,10 +578,163 @@ func LintErrorTypeAssertions(fset *token.FileSet, info *TypesInfoExt) []analysis
 			continue
 		}
 
-		lints = append(lints, analysis.Diagnostic{
+		diagnostic := analysis.Diagnostic{
 			Message: "type switch on error will fail on wrapped errors. Use errors.As to check for specific errors",
 			Pos:     typeAssert.Pos(),
-		})
+		}
+
+		// Transform type switch into a switch statement with errors.As in each case
+		// e.g., switch err.(type) { case *MyError: ... } becomes:
+		// var myError *MyError; switch { case errors.As(err, &myError): ... }
+
+		// Get the error variable being type-switched on
+		errExpr := typeAssert.X
+
+		// Determine if this is a type switch with assignment (switch e := err.(type))
+		var assignIdent *ast.Ident
+		var useShadowVar bool
+		if assignStmt, ok := typeSwitch.Assign.(*ast.AssignStmt); ok {
+			// This is a type switch with assignment like: switch e := err.(type)
+			if len(assignStmt.Lhs) == 1 {
+				if id, ok := assignStmt.Lhs[0].(*ast.Ident); ok {
+					assignIdent = id
+					useShadowVar = true
+				}
+			}
+		}
+
+		// Create variable declarations for each type
+		varDecls := []ast.Stmt{}
+
+		// Create a map of type expressions to variable names
+		typeToVar := make(map[ast.Expr]string)
+
+		// First collect all unique types from cases
+		caseTypes := []ast.Expr{}
+		for _, stmt := range typeSwitch.Body.List {
+			caseClause := stmt.(*ast.CaseClause)
+			for _, typeExpr := range caseClause.List {
+				// Skip default case (empty list)
+				if typeExpr != nil {
+					caseTypes = append(caseTypes, typeExpr)
+				}
+			}
+		}
+
+		// Create variable declarations for each type
+		for i, typeExpr := range caseTypes {
+			var varName string
+			if useShadowVar {
+				// If we have an assignment identifier, use it for all variables in a switch with assignment
+				varName = assignIdent.Name
+			} else if assignIdent != nil && i == 0 {
+				// Otherwise, if we have an assignment but not shadowing, use it for the first variable
+				varName = assignIdent.Name
```

Comment:
These two `if` assign sets the same value

maybe it's normal but it looks a bit strange, so I prefer to raise it